File: cpp/design-patterns/strategy/context/Checkout.cpp
class Checkout {
    std::unique_ptr<PaymentStrategy> strategy;
public:
    explicit Checkout(std::unique_ptr<PaymentStrategy> s)
        : strategy(std::move(s)) {}

    void setStrategy(std::unique_ptr<PaymentStrategy> s) {
        strategy = std::move(s);
    }

    double pay(double amount) {
        return strategy->charge(amount); // pure delegation
    }
};

// ── Client ──
Checkout cart(std::make_unique<CardStrategy>());
cart.pay(100);                 // 103.20
cart.setStrategy(std::make_unique<CryptoStrategy>());
cart.pay(100);                 // 101.00

File: java/design-patterns/strategy/context/Checkout.java
public class Checkout {
    private PaymentStrategy strategy;

    public Checkout(PaymentStrategy strategy) {
        this.strategy = strategy;
    }

    public void setStrategy(PaymentStrategy strategy) {
        this.strategy = strategy;
    }

    public double pay(double amount) {
        return strategy.charge(amount);   // pure delegation
    }
}

// ── Client ──
Checkout cart = new Checkout(new CardStrategy());
cart.pay(100);                 // 103.20

cart.setStrategy(new CryptoStrategy());
cart.pay(100);                 // 101.00